### ISSUE TITLE
Flutter initial

### ISSUE THREAD

**Issue by jakubprogramming** [LB31/HikingForNerds#6]
Research to flutter, how it works, how to use it and if it is suitable for the project give an introduction for all other team members.

**Comment by jakubprogramming** [comment]
As I see Flutter as kind of a direct competitor to React Native, I decided to not only analyze Flutter, but also to compare it to React Native in this process.

• React Native has a reduced development time, but Flutter offers a better usability.
• Due to the Dart Programming language, Flutter offers better performance.
• Flutter has far less resources and a smaller community.
• Both frameworks allow building Apps that look and feel like native applications, but with Flutter it is also possible to create Apps that look and feel exactly the same on both platforms. On Android Flutter by default implements Material Design. On iOS developers can choose between Material Design and Cupertino widgets.
• Flutter may also be used for developing desktop and web applications natively (early development stage), while React Native relies on third party libraries to achieve this.
• Flutter is said to be much faster than native development. Compared to React Native it may also be faster, as it does not require separate optimization for each platform (iOS/Android).
• What flutter lacks in maturity it makes up in great documentation and support by the flutter team.
• Far less libraries and tutorials are available for Flutter.
• React has been around for over 4 years, while Flutter 1.0 has officially been released
less than 1 year ago (December 2018).
• When only an Android App is needed, React Native offers no benefit. Native devel- opment is the way to go. For iOS react native may be recommended over native iOS development, as react native has better support for iOS than it does for Android.
• Cross Plattform development on Flutter requires far less configuration for each in- dividual platform. In React Native we mostly have to use different components for Android and iOS while with flutter we can use the same Widgets for both platforms in pretty much all cases.
• Flutter offers greater performance as its code is compile to native code. React Native on the other hand is not. It still has the JavaScript layer making it less performant when compared to flutter.
• Bugfixes are release much faster by the Flutter team than in React Native. Many companies using React Native in production rely on a custom fork to fix the bugs themselves.


**My Conclusion:** 
- As most of us have a fair amout of JavaScript experience, working with React Native may be easiert than learning a new programming language and Flutter from scratch
- Two of us have experience with React, so learning react native will be easier than learning Flutter. I am not sure if we really want to spend too much time on learning a completly new technology. 
- Flutter makes only sense if we really do want to develop for both platforms. If we only want to develop for one of the platforms, I think react native is the better choice. (or native development) 
- If we do want to develop for both platforms and we care greatly about offering a very close to native app user experience we should use react native
- learning flutter may be interesting and also a valuable skill to have for the future  
- there are very few references online in working with openstreetmaps and flutter. There is far more for React Native and openstreetmaps.
- When using Flutter, we may want to consider using the gpx data from openstreetmaps and importing it to google maps, as google maps has great support in flutter. 
    - https://stackoverflow.com/questions/28363479/import-openstreetmap-data-to-google-maps 
    - This may not really fit our project goal?
    - Google Maps is generally more popular among users. But what about privacy issues? Its still Google! (Prof. Weber Wulf might not approve).  


**Flutter Packages / Resources** 
https://pub.dev/packages/gpx
- Dart library for A library for loading, manipulating, and saveingGPS data in GPX format
    - only in version 0.0.3

**Comment by jakubprogramming** [comment]
**continuous text version**


Having finished my research on Flutter and React Native I now want to discuss which technology I believe is better suited for our task. As most members of our project team have a fair amount of JavaScript experience, working with React Native may be easier and quicker at first, as learning a new programming language and a new framework from scratch will be quite time consuming. Furthermore two of our team members already have experience in working with react. Considering that React Native is built upon the same principles this will be a further advantage. Flutter is a great choice for developing for Android and iOS simultaneously, as it enables us to offer almost exactly the same user experience on both platforms. Also, flutter is growing at rapid paste at the moment which could make it valuable skill to have in the future. If we only want to develop for one of these platforms though, React Native or especially native iOS/Android development are most likely the better approach. On the other hand, we may use React Native if we care greatly care about offering a native like user experience on both platforms without developing two separate native applications. As I was not able to find much information online about implementing an OpenStreetMaps application with Flutter, I believe React Native has a great advantage. There are many different third party modules for OpenStreetMaps available written in JavaScript that can easily be used in React Native. Another thing that came my mind was whether or not it is possible to export gpx-data from OSM and import it into google maps. According to a post I found on Stack Overflow [1] this may actually be possible. As Google Maps has great support in Flutter and it is generally more popular among users this may be an option for us. Though I believe that this may not exactly fit our project goal and furthermore I am concerned about potential privacy issues regarding using Google Maps.

**Comment by jakubprogramming** [comment]
**Dart Package for working with GPX Data**
As I have stated before, compared to React Native, Flutter currently offers much less third party libraries. A package that I did find though is the gpx package: https://pub.dev/ packages/gpx. According to its official web page, this library may be used for loading, manipulating and saving GPS data in GPX format. Considering that it is only in version 0.0.3 at the moment and that its documentation has some spelling mistakes I would not recommend using this package.

**Comment by jakubprogramming** [comment]
**Experimenting with Flutter**
In order to get a better understanding of the Flutter Framework I decided that it was a good idea to play around with it a little bit. To do so, I installed Flutter as well as the most recent versions of Android Studio and XCode on my laptop. I found an example project on Github (https://github.com/apptreesoftware/flutter_map) that implements a Leaflet map (https://leafletjs.com) that potentially may be used with OSM data. I cloned and installed this application on my iPhone and my Android Phone to get a some first hand user experience. The main insight I got from all of this is that the app actually does look, feel and behave pretty much identical on both platforms. Furthermore it served as a proof that it in fact is possible to implement map functionalities in Flutter without using Google Maps.
